instance/models/mixins/consul.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2018 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app model mixins - Consul
+"""
+
+from django.db import models
+
+import consul
+
+# Classes #####################################################################
+
+
+class ConsulConfigurationsMixin(models.Model):
+    """
+    A util that stores configurations in a Consul Key-Value store.
+    """
+    class Meta:
+        abstract = True
+
+    def save(self, *args, **kwargs):
+        super(ConsulConfigurationsMixin, self).save(*args, **kwargs)
+        self.update_consul_metadata()
+
+    def update_consul_metadata(self):
+        """
+        This method is going over all fields in this model to reflect
+        their values on Consul.
+        Values on consul will be fetched first, then updated only if
+        they changed in this version of the model using agent's `cas`
+        parameter (Check-And-Set). If PUT operation successfully
+        updated the field then we update the configuration version
+        number in Consul.
+
+        :return: The current version number, and the if they include
+                 changes or not.
+        """
+        agent = consul.Consul()
+        includes_changes = False
+
+        for field in self._meta.get_fields():
+            value = getattr(self, field.name)
+
+            index, _ = agent.get(field.name)
+            changed = agent.put(field.name, value, cas=index)
```

Comment:
We need to include the path `ocim/instances/{instance_id}/` in the key, since otherwise the instances would override each others data.

Reply:
Didn't catch that. Thanks @smarnach